h2/src/main/org/h2/mvstore/db/MVSortedTempResult.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2004-2018 H2 Group. Multiple-Licensed under the MPL 2.0,
+ * and the EPL 1.0 (http://h2database.com/html/license.html).
+ * Initial Developer: H2 Group
+ */
+package org.h2.mvstore.db;
+
+import java.util.BitSet;
+
+import org.h2.engine.Database;
+import org.h2.engine.Session;
+import org.h2.expression.Expression;
+import org.h2.mvstore.Cursor;
+import org.h2.mvstore.MVMap;
+import org.h2.mvstore.MVMap.Builder;
+import org.h2.result.ResultExternal;
+import org.h2.result.SortOrder;
+import org.h2.value.Value;
+import org.h2.value.ValueArray;
+
+/**
+ * Sorted temporary result.
+ */
+class MVSortedTempResult extends MVTempResult {
+
+    private final boolean distinct;
+    private final int[] indexes;
+    private final MVMap<ValueArray, Long> map;
+
+    private Cursor<ValueArray, Long> cursor;
+    private Value[] current;
+    private long valueCount;
+
+    private MVSortedTempResult(MVSortedTempResult parent) {
+        super(parent);
+        this.distinct = parent.distinct;
+        this.indexes = parent.indexes;
+        this.map = parent.map;
+        this.rowCount = parent.rowCount;
+    }
+
+    /**
+     * Creates a new sorted temporary result.
+     */
+    MVSortedTempResult(Session session, Expression[] expressions, boolean distinct, SortOrder sort) {
+        super(session);
+        this.distinct = distinct;
+        Database db = session.getDatabase();
+        int length = expressions.length;
+        int[] sortTypes = new int[length];
+        int[] indexes;
+        if (sort != null) {
+            indexes = new int[length];
+            int[] colIndex = sort.getQueryColumnIndexes();
+            int len = colIndex.length;
+            BitSet used = new BitSet();
+            for (int i = 0; i < len; i++) {
+                int idx = colIndex[i];
+                used.set(idx);
+                indexes[i] = idx;
+                sortTypes[i] = sort.getSortTypes()[i];
+            }
+            int idx = 0;
+            for (int i = len; i < length; i++) {
+                idx = used.nextClearBit(idx);
+                indexes[i] = idx;
+                idx++;
+            }
+            sameOrder: {
+                for (int i = 0; i < length; i++) {
+                    if (indexes[i] != i) {
+                        break sameOrder;
+                    }
+                }
+                indexes = null;
+            }
+        } else {
```

Comment:
same comments for this chunk of code would be nice - every time I look at this kind of code I have to mentally reverse engineer it

Reply:
I have some plans to document this logic, there are many complicated places. This pull request in its current state is mostly a preview.